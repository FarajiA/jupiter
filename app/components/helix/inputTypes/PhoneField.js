import React from 'react';
import PropTypes from 'prop-types';
import { get, toUpper } from 'lodash';
import Error from '../Error';
import PhoneInput from 'react-phone-input-2';
import 'react-phone-input-2/lib/style.css';

class PhoneField extends React.Component {
  onChange = (inputValue, countryData, event, formattedNumber) => {
    this.props.input.onChange(this.formatValue(countryData, formattedNumber));
  };

  onBlur = (event, countryData) => {
    this.props.input.onBlur(this.formatValue(countryData, event.target.value));
  };

  /** Format large data-set coming from component into a smaller data-set we can send back to redux-form */
  formatValue = (countryData, formattedValue) => {
    const format = get(countryData, 'format');
    return {
      inputValue: formattedValue.split(' ').slice(1).join(''),
      number: formattedValue.replace(/[- .()]?/g, ''),
      formattedValue,
      valid: formattedValue && format ? get(countryData, 'format').length === formattedValue.length : false,
      countryCode: toUpper(get(countryData, 'dialCode'))
    };
  };

  render() {
    const { name, label, meta, t } = this.props;
    const errorBorder = (meta.touched && meta.invalid) && ' error-border';
    return (
      <div className="InputField">
        <label htmlFor={name}>
          <span className="InputField-label">{label}</span>
        </label>
        <PhoneInput
          inputClass={`hxTextCtrl${errorBorder}`}
          country="us" // default country
          onBlur={this.onBlur}
          onChange={this.onChange}
          buttonClass={errorBorder}
          searchPlaceholder={t('common:actions.basic.search')}
          searchNotFound={t('common:search.status.notFound')}
          enableSearch
          enableLongNumbers
          disableSearchIcon
          inputProps={{
            name,
            required: true,
            autoFocus: false
          }}
        />
        <Error meta={meta} />
      </div>
    );
  }
}

PhoneField.propTypes = {
  input: PropTypes.shape({
    onChange: PropTypes.func.isRequired,
    onBlur: PropTypes.func.isRequired
  }).isRequired,
  meta: PropTypes.shape({
    touched: PropTypes.bool,
    invalid: PropTypes.bool,
    error: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.array
    ])
  }),
  name: PropTypes.string,
  label: PropTypes.string.isRequired,
  t: PropTypes.func.isRequired
};

export default PhoneField;
