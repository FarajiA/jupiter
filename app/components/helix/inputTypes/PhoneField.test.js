import React from 'react';
import renderer from 'react-test-renderer';
import PhoneField from './PhoneField';

describe('PhoneField', () => {
  let mounted;
  const defaultProps = {
    input: {
      onChange: jest.fn(),
      onBlur: jest.fn()
    },
    meta: {},
    name: 'phone name',
    id: 'fooId',
    label: 'Phone test'
  };

  beforeEach(() => {
    mounted = mount(<PhoneField {...defaultProps} />);
  });

  test('it renders with props', () => {
    const component = renderer.create(<PhoneField {...defaultProps} />).toJSON();
    expect(component).toMatchSnapshot();
  });

  test('it calls input onChange prop when onChange is called', () => {
    const spy = jest.spyOn(mounted.instance(), 'formatValue');
    const onChangeArgs = [
      true,
      1234567890,
      { iso2: 'US' },
      '(123) 456 7890'
    ];
    const expected = {
      valid: true,
      inputValue: 1234567890,
      number: '1234567890',
      countryCode: 'US'
    };
    mounted.instance().onChange(...onChangeArgs);
    expect(mounted.prop('input').onChange).toHaveBeenCalled();
    expect(spy).toHaveBeenCalled();
    expect(mounted.prop('input').onChange).toHaveBeenCalledWith(expected);
  });

  test('it calls input onBlur prop when onBlur is invoked', () => {
    const spy = jest.spyOn(mounted.instance(), 'formatValue');
    const onChangeArgs = [
      true,
      1234567890,
      { iso2: 'US' },
      '(123) 456 7890'
    ];
    const expected = {
      valid: true,
      inputValue: 1234567890,
      number: '1234567890',
      countryCode: 'US'
    };
    mounted.instance().onBlur(...onChangeArgs);
    expect(mounted.prop('input').onBlur).toHaveBeenCalled();
    expect(spy).toHaveBeenCalled();
    expect(mounted.prop('input').onBlur).toHaveBeenCalledWith(expected);
  });

  test('format value formats and returns the args', () => {
    const onChangeArgs = [
      true,
      1234567890,
      { iso2: 'US' },
      '(123) 456 7890'
    ];
    const expected = {
      valid: true,
      inputValue: 1234567890,
      number: '1234567890',
      countryCode: 'US'
    };
    const returned = mounted.instance().formatValue(...onChangeArgs);
    expect(returned).toEqual(expected);
  });
});
