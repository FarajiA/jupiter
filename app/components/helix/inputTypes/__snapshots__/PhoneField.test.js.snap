// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneField PhoneField renders correct props 1`] = `
<div
  className="InputField"
>
  <label
    htmlFor="phone"
  >
    <span
      className="InputField-label"
    >
      Phone Number
    </span>
  </label>
  <div
    className=" react-tel-input"
    onKeyDown={[Function]}
    style={Object {}}
  >
    <div
      className="special-label"
    >
      Phone
    </div>
    <input
      autoFocus={false}
      className="hxTextCtrl form-control"
      disabled={false}
      name="phone"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onCopy={[Function]}
      onDoubleClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="1 (702) 123-4567"
      required={true}
      style={Object {}}
      type="tel"
      value="+44"
    />
    <div
      className="false flag-dropdown"
      style={Object {}}
    >
      <div
        aria-haspopup="listbox"
        className="selected-flag"
        onClick={[Function]}
        role="button"
        tabIndex="0"
        title="United Kingdom: + 44"
      >
        <div
          className="flag gb"
        >
          <div
            className="arrow"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="error"
  />
</div>
`;
